File: web/src/lib/crypto/symmetric.ts
import nacl from 'tweetnacl';
import { decodeUTF8, encodeUTF8, encodeBase64, decodeBase64 } from 'tweetnacl-util';
import type { EncryptedPayload } from '@/types/chat';

// ─── 바이너리 (이미지 등) 대칭키 암복호화 ───

/**
 * 바이너리 대칭키 암호화 (nacl.secretbox - XSalsa20-Poly1305)
 * 이미지 등 Uint8Array 데이터용. Base64 EncryptedPayload 반환.
 */
export function encryptBinary(
  data: Uint8Array,
  symmetricKey: Uint8Array
): EncryptedPayload {
  const nonce = nacl.randomBytes(nacl.secretbox.nonceLength);
  const ciphertext = nacl.secretbox(data, nonce, symmetricKey);

  if (!ciphertext) {
    throw new Error('Binary encryption failed');
  }

  return {
    ciphertext: encodeBase64(ciphertext),
    nonce: encodeBase64(nonce),
  };
}

/**
 * 바이너리 대칭키 복호화 (raw Uint8Array 입출력)
 * API Route에서 raw 바이너리를 받아 base64 변환 없이 직접 복호화.
 */
export function decryptBinaryRaw(
  ciphertext: Uint8Array,
  nonce: Uint8Array,
  symmetricKey: Uint8Array
): Uint8Array | null {
  return nacl.secretbox.open(ciphertext, nonce, symmetricKey);
}

// ─── 문자열 대칭키 암복호화 ───

/**
 * 대칭키 암호화 (nacl.secretbox - XSalsa20-Poly1305)
 *
 * 게시판 등 1:N 공유 대칭키 환경용.
 * deriveKeysFromPassword()의 encryptionSeed(32B)를 키로 사용.
 */
export function encryptSymmetric(
  plaintext: string,
  symmetricKey: Uint8Array
): EncryptedPayload {
  const nonce = nacl.randomBytes(nacl.secretbox.nonceLength);
  const messageBytes = decodeUTF8(plaintext);
  const ciphertext = nacl.secretbox(messageBytes, nonce, symmetricKey);

  if (!ciphertext) {
    throw new Error('Symmetric encryption failed');
  }

  return {
    ciphertext: encodeBase64(ciphertext),
    nonce: encodeBase64(nonce),
  };
}

/**
 * 대칭키 복호화 (nacl.secretbox.open - XSalsa20-Poly1305)
 */
export function decryptSymmetric(
  payload: EncryptedPayload,
  symmetricKey: Uint8Array
): string | null {
  const ciphertext = decodeBase64(payload.ciphertext);
  const nonce = decodeBase64(payload.nonce);
  const decrypted = nacl.secretbox.open(ciphertext, nonce, symmetricKey);

  if (!decrypted) {
    return null;
  }

  return encodeUTF8(decrypted);
}

File: mobile/lib/core/crypto/symmetric.dart
import 'dart:convert';

import 'package:pinenacl/x25519.dart';

import 'encrypt.dart';

/// 대칭키 암호화 (web: nacl.secretbox)
/// 게시판 등 N:N 공유 환경용
/// pinenacl SecretBox는 crypto_box_afternm 사용 → nacl.secretbox과 동일 구현
EncryptedPayload encryptSymmetric(String plaintext, Uint8List symmetricKey) {
  final box = SecretBox(symmetricKey);
  final encrypted = box.encrypt(Uint8List.fromList(utf8.encode(plaintext)));

  return EncryptedPayload(
    ciphertext: base64Encode(Uint8List.fromList(encrypted.cipherText)),
    nonce: base64Encode(Uint8List.fromList(encrypted.nonce)),
  );
}

/// 대칭키 복호화 (web: nacl.secretbox.open)
String? decryptSymmetric(EncryptedPayload payload, Uint8List symmetricKey) {
  try {
    final ciphertext = base64Decode(payload.ciphertext);
    final nonce = base64Decode(payload.nonce);

    final box = SecretBox(symmetricKey);
    final encrypted = EncryptedMessage(
      nonce: Uint8List.fromList(nonce),
      cipherText: Uint8List.fromList(ciphertext),
    );
    final plaintext = box.decrypt(encrypted);

    return utf8.decode(plaintext);
  } catch (_) {
    return null;
  }
}

/// 바이너리 대칭키 암호화 (이미지 등)
EncryptedPayload encryptBinary(Uint8List data, Uint8List symmetricKey) {
  final box = SecretBox(symmetricKey);
  final encrypted = box.encrypt(data);

  return EncryptedPayload(
    ciphertext: base64Encode(Uint8List.fromList(encrypted.cipherText)),
    nonce: base64Encode(Uint8List.fromList(encrypted.nonce)),
  );
}

/// 바이너리 대칭키 복호화 (raw Uint8List 반환)
Uint8List? decryptBinaryRaw(
  Uint8List ciphertext,
  Uint8List nonce,
  Uint8List symmetricKey,
) {
  try {
    final box = SecretBox(symmetricKey);
    final encrypted = EncryptedMessage(
      nonce: nonce,
      cipherText: ciphertext,
    );
    return Uint8List.fromList(box.decrypt(encrypted));
  } catch (_) {
    return null;
  }
}
